src/DataGrid.tsx

```diff
@@ -368,43 +550,130 @@ function DataGrid<R, K extends keyof R, SR>({
     }
   }
 
-  function getViewportRows() {
-    const rowElements = [];
+  function navigate(key: string, shiftKey: boolean) {
+    let nextPosition = getNextPosition(key, shiftKey, selectedPosition);
+    let mode = cellNavigationMode;
+    if (key === 'Tab') {
+      // If we are in a position to leave the grid, stop editing but stay in that cell
+      if (canExitGrid({ shiftKey, cellNavigationMode, columns, rowsCount: rows.length, selectedPosition })) {
+        // Reset the selected position before exiting
+        setSelectedPosition({ idx: -1, rowIdx: -1, mode: 'SELECT' });
+        return;
+      }
 
-    for (let rowIdx = rowOverscanStartIdx; rowIdx <= rowOverscanEndIdx; rowIdx++) {
-      const row = rows[rowIdx];
-      let key: string | number = rowIdx;
-      let isRowSelected = false;
-      if (rowKey !== undefined) {
-        const rowId = row[rowKey];
-        isRowSelected = selectedRows?.has(rowId) ?? false;
-        if (typeof rowId === 'string' || typeof rowId === 'number') {
-          key = rowId;
+      mode = cellNavigationMode === CellNavigationMode.NONE
+        ? CellNavigationMode.CHANGE_ROW
+        : cellNavigationMode;
+    }
+
+    nextPosition = getNextSelectedCellPosition<R, SR>({
+      columns,
+      rowsCount: rows.length,
+      cellNavigationMode: mode,
+      nextPosition
+    });
+
+    selectCell(nextPosition);
+  }
+
+  function getDraggedOverCellIdx(currentRowIdx: number): number | undefined {
+    if (draggedOverRowIdx === undefined) return;
+    const { rowIdx } = selectedPosition;
+
+    const isDraggedOver = rowIdx < draggedOverRowIdx
+      ? rowIdx < currentRowIdx && currentRowIdx <= draggedOverRowIdx
+      : rowIdx > currentRowIdx && currentRowIdx >= draggedOverRowIdx;
+
+    return isDraggedOver ? selectedPosition.idx : undefined;
+  }
+
+  function getSelectedCellProps(rowIdx: number): SelectedCellProps | undefined {
```

Comment:
This function is called on each render, and then passed the return result into the `Row`. I am wondering if we shall add memorization to the output to prevent unnecessary Row re-render.

Reply:
It is only passed to the row containing the selected cell so we only rerender 2 rows max. The performance is good and it is tricky to memoize this function. I think it is fine for now